---
title: "Exploration of the Ec3.2.1.37 Bacterial Protein Clusters From CAZy"
author: "Emma E. M. Hobbs"
date: "Feb 2022"
output: 
  bookdown::html_document2:
    toc: true
    toc_float:
      toc_collapsed: false
    number_sections: true
    theme: lumen
---

# Introduction

This notebook contains all code for parsing and analysis the protein clusters generated using [`MMSeq2`](https://github.com/soedinglab/MMseqs2).

```{r setup, include=FALSE}
# imports
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
library('kableExtra')
library(data.table)
library('ggplot2')
library("dplyr")
library("GGally")
library(knitr)
library(tidyverse)
library(RColorBrewer)
library(cowplot)
library(viridis)
library(plotly)
```

```{r loadData, include=FALSE}
# import the data
rep_seq_blastp <- read.csv("cluster_rep_blastp.tsv", header=FALSE, sep="\t")

AGE22437_1_blastp <- read.csv("cluster_sequences/AGE22437_1_blastp.tsv", header=FALSE, sep="\t")
CBK6950_1_blastp <- read.csv("cluster_sequences/CBK6950_1_blastp.tsv", header=FALSE, sep="\t")
CDG29680_1_blastp <- read.csv("cluster_sequences/CDG29680_1_blastp.tsv", header=FALSE, sep="\t")
QJR11213_1_blastp <- read.csv("cluster_sequences/QJR11213_1_blastp.tsv", header=FALSE, sep="\t")
AGE22437_1_CDG29680_1_blastp <- read.csv("cluster_sequences/AGE22437_1_CDG29680_1_blastp.tsv", header=FALSE, sep="\t")
all_clusters_blastp <- read.csv("cluster_sequences/all_clusters_blastp.tsv", header=FALSE, sep="\t")
remaining_fam_seqs_blastp <- read.csv("cluster_sequences/remaining_fam_seqs_blastp.tsv", header=FALSE, sep="\t")
```

# Method

To explore the degree of sequence divergencen and similarity in each of the clusters created by `MMSeq2`, all-versus-all BLASTP analysis was performed. In every instance, the script `run_blastp.py` from the Python package [`pyrewton` DOI:10.5281/zenodo.3876218)](https://github.com/HobnobMancer/pyrewton) was used to run the BLASTP all-versus-all analysis.

The table compiled by BLASTP contains the following columns:
- qseqid
- sseqid
- pident
- length
- mismatch
- gapopen
- qstart
- qend
- sstart
- send
- evalue
- bitscore

In order to compare each of the pair-wise-alignment we need to calculate the Blast Score Ratio (SCR) to normalise for length. This was first presented by [Rasko _et al._, 2005](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-6-2).  

The bitscore reported by BLAST is the sum of the qualities of the aligned symbols over the whole alignment. This is an accurate measure of the alignment strength, but long sequences tend to have higher bitscores than short sequences, even when the matches are of about the same quality. To correct for this length effect, we can calculate a normalised bitscore where:

normalised bitscore = bitscore / query length  

```{r blastpAnalysis, include=FALSE}
parse_data <- function(df){
  # change column names
  setnames(
    df,
    old = c('V1','V2','V3','V4','V5','V6','V7','V8','V9'),
    new = c('qseqid','sseqid','pident','cov','qlen','slen','alen','bitscore','evalue')
  )
  # calculate the BSR
  df$bsr = df$bitscore/df$qlen
  
  return(df)
}

get_interactive_plot <- function(df) {
  data <- df %>%
    mutate(text = paste0("Query seq ID: ", qseqid, "\n", "Subject seq ID: ", sseqid, "\n", "BSR: ",round(bsr,3)))
  
  p <- ggplot(data, aes(qseqid, sseqid, fill= bsr, text=text)) + 
    geom_tile() +
    scale_fill_viridis(discrete=FALSE) +
    xlab("Query seq ID") + 
    ylab("Subject seq ID") +
    theme(axis.text.x = element_text(angle=90, vjust=.5, hjust=1))
  p <- ggplotly(p, tooltip="text")
  return(p)
}

get_static_plot <- function(df) {
  p <- ggplot(df, aes(qseqid, sseqid, fill= bsr)) + 
    geom_tile() +
    scale_fill_viridis(discrete=FALSE) +
    xlab("Query seq ID") + 
    ylab("Subject seq ID") +
    theme(axis.text.x = element_text(angle=90, vjust=.5, hjust=1))
  
  return(p)
}
```


# Preliminary comparison between clusters

A representative sequence from each of the 4 largest clusters compiled using `MMSeq2` (with a percentage identity and coverage of cut-off of 70%) was extracted from a local CAZyme database using `cazy_webscraper`.

Each representative sequence was identified by using the GenBank accession assigned as the name of the cluster by `MMSeq2`.

Table \@ref(tab:blastpDF) presents the raw output from BLASTP as well as the BSR from the all-verus-all BLASTP analysis of the representative sequences from the 4 largest clusters compiled by `MMSeq2`.

```{r blastpDF, echo=FALSE, fig.cap="Summary statistics of CAZyme classifiers performances of binary CAZyme/non-CAZyme prediction. Data collected is the mean of call calculated statistical parameters across all test sets, plus and minus the standard devliation. All figures are rounded to 4 decimal places."}
rep_seq_blastp.parsed <- parse_data(rep_seq_blastp)

kable(rep_seq_blastp.parsed, caption="Output from BLASTP all-vs-all analysis of the representative sequences from the 4 larges protein clusters", align='c') %>% kable_styling(full_width = F)
```

Figure \@ref(fig:blastpFIG) is an interactive plot presenting the BSR from the all-verus-all BLASTP analysis of the representative sequences from the 4 largest clusters compiled by `MMSeq2`.

To view the specific BSR for each comparison, hover over the plot and a tooltip will appear and will present the GenBank accessions of the corresponding proteins as well as the specific BSR value (to 3dp).

```{r blastpFIG, echo=FALSE, fig.cap="One-dimensional scatter plot of specificity scores of CAZyme and non-CAZyme predictions per test set, overlaying box plot of standard deviation."}
blastp.plot.interactive <- get_interactive_plot(rep_seq_blastp.parsed)

# htmlwidgets::saveWidget(blastp.plot.interactive, "cluster_rep_results.html")

blastp.plot.static <- get_static_plot(rep_seq_blastp.parsed)
# pdf(blastp.plot.static, file="pl_cluster_analysis/xp_003719456/xp_003719456.static.pdf", width=10, height=8)

blastp.plot.interactive
```

The cluster CBK69950.1 contained 33 protein sequences, QJR11213.1	28 protein sequences, CDG29680.1 17 protein sequences, and AGE22437.1	contaiend 13 protein sequences.  

The BSR infer that the two smaller clusters (CDG29680.1 and AGE22437.1) could potentially be combined to create a large cluster. The proteins in this new combined clusters could then be aligned to create an multisequence alignment (MSA) of functionally relevant proteins for molecular modeling. Conversely, the BSR inferred the two larger clusters should be kept separate ( and not combined) to create a high quality MSA (i.e. minimal gaps and readily identifiable consensus sequences) of the proteins within each cluster.

# Sequence divergence in individual clusters

To explore the sequence divergence in each of the 4 largest clusters created by `MMSeq2`, for each cluster the protein sequences were retrieved using `cazy_webscraper` and were written to a FASTA file. The protein sequences were then compared to one another in a BLASTP all-versus-all analysis.

## AGE22437.1

Figure \@ref(fig:AGE22437) presents the heatmap of the all-versus-all BLASTP analysis of the 13 protein sequences in the AGE22437.1 cluster. The sequence BAM46395.1 has the least sequence similarity to all other members of the cluster. Overall, the cluster has a relatively similar sequence as inferred from the BSR.

```{r AGE22437, echo=FALSE, fig.cap="Heatmap of all-versus-all BLASTP of the AGE22437.1 cluster"}
AGE22437_1_blastp.parse <- parse_data(AGE22437_1_blastp)
AGE22437_1.blastp.plot.interactive <- get_interactive_plot(AGE22437_1_blastp.parse)
AGE22437_1.blastp.plot.interactive
```

## CBK6950.1

Figure \@ref(fig:CBK6950) presents the heatmap of the all-versus-all BLASTP analysis of the 33 protein sequences in the CBK6950.1 cluster. Except for 2 instances, every pair-wise alignment produced a BSR of greater than 1.3 with many with a BRS greater than 1.5. This infers a strong sequence similarity across all proteins in the cluster. However,  
- CBL10126.1 against VEI47713.1
- CBL13352.1 against VEI47713.1
- VEI47713.1 against CBL10126.1
- VEI47713.1 against CBL13352.1
produced BRS of 0.04. 

```{r CBK6950, echo=FALSE, fig.cap="Heatmap of all-versus-all BLASTP of the CBK6950.1 cluster"}
CBK6950_1_blastp.parse <- parse_data(CBK6950_1_blastp)
CBK6950_1_blastp.plot.interactive <- get_interactive_plot(CBK6950_1_blastp.parse)
CBK6950_1_blastp.plot.interactive
```

## CDG29680.1

Figure \@ref(fig:CDG29680) presents the heatmap of the all-versus-all BLASTP analysis of the 17 protein sequences in the CDG29680.1 cluster. As with CBK6950.1, overall all proteins in the cluster have a similar protein sequence to one another as inferred from the high (greater than 1.4) BSR. However, there are a few instances of pairwise alignments with BSRs of less than 0.5.

```{r CDG29680, echo=FALSE, fig.cap="Heatmap of all-versus-all BLASTP of the CDG29680.1 cluster"}
CDG29680_1_blastp.parse <- parse_data(CDG29680_1_blastp)
CDG29680_1_blastp.plot.interactive <- get_interactive_plot(CDG29680_1_blastp.parse)
CDG29680_1_blastp.plot.interactive
```

## QJR11213.1

Figure \@ref(fig:QJR11213) presents the heatmap of the all-versus-all BLASTP analysis of the 28 protein sequences in the QJR11213.1 cluster. The heatmap shows clusters of proteins with high sequence similarity, as inferred from the high (greater than 1.8) BSR. AK102785.1, QJR11213.1 and QJR15254.1 both relatively lower BSRs of approximately 1.4 against all other proteins in the cluster. However, a BSR of 1.4 is still a relatively high BSR, therefore, there is relatively little protein sequence variation in this cluster.

```{r QJR11213, echo=FALSE, fig.cap="Heatmap of all-versus-all BLASTP of the QJR11213.1 cluster"}
QJR11213_1_blastp.parse <- parse_data(QJR11213_1_blastp)
QJR11213_1_blastp.plot.interactive <- get_interactive_plot(QJR11213_1_blastp.parse)
QJR11213_1_blastp.plot.interactive
```

# Combining clusters

## AGE22437.1 and CDG29680.1

Figure \@ref(fig:AGE22437) inferred that the clusters AGE22437.1 and CDG29680.1 could potentially be combined. To further explore this the a BLASTP all-versus-all analysis of all proteins in both clusters was performed. The BSRs from this analysis are presented in figure \@ref(fig:AGE22437CDG29680}, and the plot demonstrates that the vast majoirty of pair-wise alignements produced a BSR of greater than 1.4. This infers relatively sequence divergence between all proteins across the two clusters, and therefore, the two clusters can be combined.

```{r AGE22437CDG29680, echo=FALSE, fig.cap="Heatmap of all-versus-all BLASTP of the proteins from both AGE22437.1 and CDG29680.1 clusters"}
AGE22437_1_CDG29680_1_blastp.parse <- parse_data(AGE22437_1_CDG29680_1_blastp)
AGE22437_1_CDG29680_1_blastp.plot.interactive <- get_interactive_plot(AGE22437_1_CDG29680_1_blastp.parse)
AGE22437_1_CDG29680_1_blastp.plot.interactive
```

## Sequence divergence across all 4 clusters

To explore the possibility of creating a MSA with a consensus across the majority of proteins when aligning proteins from all 4 clusters (AGE22437.1, CBK6950.1, CDG29680.1 and QJR11213.1), a all-versus-all BLASTP analysis of all proteins in all 4 clusters was performed. The BSR of all pairwise alignments are presented in figure \@ref(allClustersBlastp). Clear clusters of proteins with high BSRs (greater than 1.5) were observed, and these clusters included proteins from across multiple clusters. The majority of pairwise alignments scores a BSR of 1 and greater. Consequently, it was hypothesised that all 4 clusters could be combined to generate a MSA of functionally relevant proteins for molecular modeling.

```{r allClustersBlastp, echo=FALSE, fig.cap="Heatmap of all-versus-all BLASTP of the proteins from both AGE22437.1, CBK6950.1, CDG29680.1 and QJR11213.1 clusters"}
all_clusters_blastp.parse <- parse_data(all_clusters_blastp)
all_clusters_blastp.plot.interactive <- get_interactive_plot(all_clusters_blastp.parse)
all_clusters_blastp.plot.interactive
```

# Remaining family proteins against the clusters

The total number of proteins in across all 4 clusters was 91. This included 0 proteins with PDB accessions listed in UniProt. The following SQL command was used to retrive the results:

```{r, eval=FALSE}
WITH Ec_Query (ec_gbk_acc) AS (
	SELECT DISTINCT Genbanks.genbank_accession
	FROM Genbanks
	INNER JOIN Genbanks_Ecs ON Genbanks.genbank_id = Genbanks_Ecs.genbank_id
	INNER JOIN Ecs ON Genbanks_Ecs.ec_id = Ecs.ec_id
	WHERE Ecs.ec_number = '3.2.1.37'
)
SELECT DISTINCT Genbanks.genbank_accession, Pdbs.pdb_accession
FROM Genbanks
INNER JOIN Pdbs ON Genbanks.genbank_id = Pdbs.genbank_id
LEFT JOIN Ec_Query ON Genbanks.genbank_accession = Ec_Query.ec_gbk_acc
WHERE (Genbanks.genbank_accession IN Ec_Query)
```

To further expand the pool of potentially functionally relevant proteins, the proteins from the CAZy families of interest (listed below) which were not included in the clusters were BLASTP queries against the members of the 4 clusters.

**Glycoside Hydrolase CAZy families of interest:**
- GH1
- GH2
- GH3
- GH11
- GH26
- GH30
- GH43
- GH51
- GH52
- GH54
- GH116
- GH120

A percentage identity cut-off of 70% was applied, similar to the clustering cut-off applied by `MMSeq2`. The BSR was the calculated and a heatmap plotted, which is presented in figure \@ref{fig:remainingFamsBlastp}.

```{r remainingFamsBlastp, echo=FALSE, fig.cap="Heatmap of BSR between GH proteins with a percentage identitity of greater than or equal to 70% against proteins from the 4 largest `MMSeq2` protein clusters of proteins annotated with the EC number 3.2.1.37"}
remaining_fam_seqs_blastp.parse <- parse_data(remaining_fam_seqs_blastp)

# filter for pair-wise alignments of >= 70%
subset_blastp <- remaining_fam_seqs_blastp.parse[which(remaining_fam_seqs_blastp.parse$pident >= 70), ]

data <- subset_blastp %>%
  mutate(text = paste0("Query seq ID: ", qseqid, "\n", "Subject seq ID: ", sseqid, "\n", "BSR: ",round(bsr,3)))

subset_blastp.p <- ggplot(data, aes(qseqid, sseqid, fill= bsr, text=text)) + 
  geom_tile() +
  scale_fill_viridis(discrete=FALSE) +
  xlab("Query seq ID") + 
  ylab("Subject seq ID") +
  theme(axis.text.x = element_text(angle=90, vjust=.5, hjust=1))
subset_blastp.p <- ggplotly(subset_blastp.p, tooltip="text")
subset_blastp.p
```

After filtering for a percentage identity of greater than or equal to 70%, the BSR was plotted for the remaining proteins. The majority of the remaining proteins produced a BSR of less than 1. However, the following query proteins produced a high BSR (of 1.5-1.9) against the the majority of the proteins in the protein clusters:
- QFT14636.1
- VDR25565.1
- VDZ67166.1
- VED48126.1
- VEI47712.1
- VTO17894.1
- VTP66272.1
- VTP66274.1

Where no data is presented in the figure \@ref{fig:remainingFamsBlastp} the specific pairwise alignment between the query and subject protein sequences did not have a percentage identity of greater than or equal to 70%. In order to produce a MSA with a clear consensus sequence for molecular modeling, proteins including in the MSA need to have have sequence identity with the majority of other proteins in the MSA. Therefore, proteins from the BLASTP search of the remaining proteins from the CAZy families of interest against the 4 largest clusters from `MMSeq2` which did **not** have a percentage identity of greater than or equal to 70% against the majority of proteins from the 4 largest `MMSeq2` clusters were **not** added to the collection of proteins to be included in the MSA.

The proteins which did have a percentage identity of greater than or equal to 70% against high BSR (of 1.5-1.9) against the the majority of the proteins in the protein clusters the majority of proteins from the 4 largest `MMSeq2` clusters were included in the collection of proteins to be included in the MSA. These were the 8 proteins with a high BSR (of 1.5-1.9) against the the majority of the proteins in the protein clusters. None of these proteins were linked to any PDB accessions in UniProt.
